File: src/Stopwatch_TEST.cc
/*
 * Copyright (C) 2018 Open Source Robotics Foundation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/

#include <gtest/gtest.h>
#include <thread>

#include "ignition/math/Stopwatch.hh"

using namespace ignition;

/////////////////////////////////////////////////
// Helper function that runs a few tests
void runTimer(math::Stopwatch &_time)
{
  // Windows uses a system_clock for std::this_thread::sleep_for. This can
  // cause incorrect sleep durations. So, we add some room for error on
  // windows.
  std::chrono::duration<int, std::milli> handleSteadyClock =
    std::chrono::milliseconds(0);
#ifdef _WIN32
  handleSteadyClock += std::chrono::milliseconds(100);
#endif

  // Start the timer
  EXPECT_TRUE(_time.Start());
  // The timer should be running.
  EXPECT_TRUE(_time.Running());
  // The start time should be greater than the stop time.
  EXPECT_GT(_time.StartTime(), _time.StopTime());
  // The elapsed stop time should still be zero.
  EXPECT_EQ(ignition::math::clock::duration::zero(),
            _time.ElapsedStopTime());

  // Wait for some time...
  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
  // Now the elapsed time should be greater than or equal to the time slept.
  EXPECT_GE(_time.ElapsedRunTime() + handleSteadyClock,
      std::chrono::milliseconds(1000));

  // Stop the timer.
  EXPECT_TRUE(_time.Stop());
  // The timer should not be running.
  EXPECT_FALSE(_time.Running());
  // The stop time should be greater than the start time.
  EXPECT_GT(_time.StopTime(), _time.StartTime());
  // The elapsed time should still be greater than the time slept.
  EXPECT_GE(_time.ElapsedRunTime() + handleSteadyClock,
      std::chrono::milliseconds(1000));

  // Save the elapsed time.
  auto elapsedTime = _time.ElapsedRunTime();

  // The timer is now stopped, let's sleep some more.
  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
  // The elapsed stop time should be greater than or equal to the time
  // slept.
  EXPECT_GE(_time.ElapsedStopTime() + handleSteadyClock,
      std::chrono::milliseconds(1000));
  // The elapsed time should be the same.
  EXPECT_EQ(elapsedTime, _time.ElapsedRunTime());

  // Start the timer again.
  EXPECT_TRUE(_time.Start());
  // Store the elapsed stop time.
  auto elapsedStopTime = _time.ElapsedStopTime();
  // The timer should be running.
  EXPECT_TRUE(_time.Running());
  // Sleep for some time.
  std::this_thread::sleep_for(std::chrono::milliseconds(1000));
  // The elapsed stop time should remain the same
  EXPECT_EQ(elapsedStopTime, _time.ElapsedStopTime());
  // The elapsed time should be greater than the previous elapsed time.
  EXPECT_GT(_time.ElapsedRunTime(), elapsedTime);
  // The elapsed time should be greater than or equal to the the previous
  // two sleep times.
  EXPECT_GE(_time.ElapsedRunTime() + handleSteadyClock,
      std::chrono::milliseconds(2000));
}

/////////////////////////////////////////////////
TEST(Stopwatch, Constructor)
{
  math::Stopwatch watch;

  EXPECT_FALSE(watch.Running());
  EXPECT_EQ(watch.StopTime(), watch.StartTime());
  EXPECT_EQ(ignition::math::clock::duration::zero(), watch.ElapsedRunTime());
  EXPECT_EQ(ignition::math::clock::duration::zero(), watch.ElapsedStopTime());

  runTimer(watch);

  math::Stopwatch watch2(watch);
  EXPECT_EQ(watch, watch2);

  math::Stopwatch watch3(std::move(watch2));
  EXPECT_EQ(watch, watch3);
}

/////////////////////////////////////////////////
TEST(Stopwatch, EqualOperator)
{
  math::Stopwatch watch;
  math::Stopwatch watch2;
  math::Stopwatch watch3;
  EXPECT_EQ(watch, watch2);
  EXPECT_EQ(watch, watch3);

  runTimer(watch);
  runTimer(watch2);
  runTimer(watch3);

  EXPECT_NE(watch, watch2);
  EXPECT_NE(watch, watch3);

  watch2 = watch;
  EXPECT_EQ(watch, watch2);

  watch3 = std::move(watch2);
  EXPECT_EQ(watch, watch3);
}

/////////////////////////////////////////////////
TEST(Stopwatch, StartStopReset)
{
  math::Stopwatch watch;

  runTimer(watch);

  watch.Reset();

  EXPECT_FALSE(watch.Running());
  EXPECT_EQ(watch.StopTime(), watch.StartTime());
  EXPECT_EQ(ignition::math::clock::duration::zero(), watch.ElapsedRunTime());
  EXPECT_EQ(ignition::math::clock::duration::zero(), watch.ElapsedStopTime());

  runTimer(watch);

  EXPECT_TRUE(watch.Running());

  watch.Start(true);
  EXPECT_TRUE(watch.Running());
  EXPECT_LT(watch.StopTime(), watch.StartTime());
  EXPECT_NE(ignition::math::clock::duration::zero(), watch.ElapsedRunTime());
  EXPECT_EQ(ignition::math::clock::duration::zero(), watch.ElapsedStopTime());
}

/////////////////////////////////////////////////
TEST(Stopwatch, FailStartStop)
{
  math::Stopwatch watch;

  // Can't stop while not running
  EXPECT_FALSE(watch.Stop());
  EXPECT_FALSE(watch.Running());

  // Can start while not running
  EXPECT_TRUE(watch.Start());
  EXPECT_TRUE(watch.Running());

  // Can't start while running
  EXPECT_FALSE(watch.Start());
  EXPECT_TRUE(watch.Running());

  // Can stop while running
  EXPECT_TRUE(watch.Stop());
  EXPECT_FALSE(watch.Running());

  // Can start while not running
  EXPECT_TRUE(watch.Start());
  EXPECT_TRUE(watch.Running());
}

File: src/python_pybind11/test/StopWatch_TEST.py
# Copyright (C) 2021 Open Source Robotics Foundation
#
# Licensed under the Apache License, Version 2.0 (the "License")
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#       http:#www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import time
import unittest
from datetime import datetime, timedelta

from ignition.math import Stopwatch


class TestBox(unittest.TestCase):
    # Helper function that runs a few tests
    def runTimer(self, _time):
        handleSteadyClock = timedelta(0)

        # Start the timer
        self.assertTrue(_time.start())
        # The timer should be running.
        self.assertTrue(_time.running())
        # The start time should be greater than the stop time.
        self.assertGreater(_time.start_time(), _time.stop_time())
        # The elapsed stop time should still be zero.
        self.assertEqual(timedelta(0), _time.elapsed_stop_time())

        # Wait for some time...
        time.sleep(1)
        # Now the elapsed time should be greater than or equal to the time slept.
        self.assertGreaterEqual(_time.elapsed_run_time() + handleSteadyClock, timedelta(seconds=1))

        # Stop the timer.
        self.assertTrue(_time.stop())
        # The timer should not be running.
        self.assertFalse(_time.running())
        # The stop time should be greater than the start time.
        self.assertGreater(_time.stop_time(), _time.start_time())
        # The elapsed time should still be greater than the time slept.
        self.assertGreaterEqual(_time.elapsed_run_time() + handleSteadyClock, timedelta(seconds=1))

        # Save the elapsed time.
        elapsedTime = _time.elapsed_run_time()

        # The timer is now stopped, let's sleep some more.
        time.sleep(1)
        # The elapsed stop time should be greater than or equal to the time
        # slept.
        self.assertGreaterEqual(_time.elapsed_stop_time() + handleSteadyClock, timedelta(seconds=1))
        # The elapsed time should be the same.
        self.assertEqual(elapsedTime, _time.elapsed_run_time())

        # Start the timer again.
        self.assertTrue(_time.start())
        # Store the elapsed stop time.
        elapsedStopTime = _time.elapsed_stop_time()
        # The timer should be running.
        self.assertTrue(_time.running())
        # Sleep for some time.
        time.sleep(1)
        # The elapsed stop time should remain the same
        self.assertEqual(elapsedStopTime, _time.elapsed_stop_time())
        # The elapsed time should be greater than the previous elapsed time.
        self.assertGreater(_time.elapsed_run_time(), elapsedTime)
        # The elapsed time should be greater than or equal to the the previous
        # two sleep times.
        self.assertGreaterEqual(_time.elapsed_run_time() + handleSteadyClock, timedelta(seconds=2))

    def test_constructor(self):
        watch = Stopwatch()

        self.assertFalse(watch.running())
        self.assertEqual(watch.stop_time(), watch.start_time())
        self.assertEqual(timedelta(0), watch.elapsed_run_time())
        self.assertEqual(timedelta(0), watch.elapsed_stop_time())

        self.runTimer(watch)

        watch2 = watch
        self.assertEqual(watch, watch2)

        watch3 = watch2
        self.assertEqual(watch, watch3)

    def test_equal_operator(self):
        watch = Stopwatch()
        watch2 = Stopwatch()
        watch3 = Stopwatch()
        self.assertEqual(watch, watch2)
        self.assertEqual(watch, watch3)

        self.runTimer(watch)
        self.runTimer(watch2)
        self.runTimer(watch3)

        self.assertNotEqual(watch, watch2)
        self.assertNotEqual(watch, watch3)

        watch2 = watch
        self.assertEqual(watch, watch2)

        watch3 = watch2
        self.assertEqual(watch, watch3)

    def test_start_stop_reset(self):
        watch = Stopwatch()

        self.runTimer(watch)

        watch.reset()

        self.assertFalse(watch.running())
        self.assertEqual(watch.stop_time(), watch.start_time())
        self.assertEqual(timedelta(0), watch.elapsed_run_time())
        self.assertEqual(timedelta(0), watch.elapsed_stop_time())

        self.runTimer(watch)

        self.assertTrue(watch.running())

        watch.start(True)
        self.assertTrue(watch.running())
        self.assertLess(watch.stop_time(), watch.start_time())
        self.assertNotEqual(timedelta(0), watch.elapsed_run_time())
        self.assertEqual(timedelta(0), watch.elapsed_stop_time())

    def test_fail_start_stop(self):
        watch = Stopwatch()

        # Can't stop while not running
        self.assertFalse(watch.stop())
        self.assertFalse(watch.running())

        # Can start while not running
        self.assertTrue(watch.start())
        self.assertTrue(watch.running())

        # Can't start while running
        self.assertFalse(watch.start())
        self.assertTrue(watch.running())

        # Can stop while running
        self.assertTrue(watch.stop())
        self.assertFalse(watch.running())

        # Can start while not running
        self.assertTrue(watch.start())
        self.assertTrue(watch.running())


if __name__ == '__main__':
    unittest.main()
